README.md
Bug-Reporter7
=============
Bug Reporter7 is a web based application to help you and your team keep your work organized by moving Bug Reporting into one centralized system.
Driven by the powerful [Framework7](http://github.com/nolimits4web/Framework7), Bug Reporter7 allows Users and Developers to advance development of their projects by submitting Bugs, which other Developers can try to fix.

Key Features
============
**NOTE: This is the Demo of Bug Reporter7! There is only a limited set of features that can be used right now.**

* **Submit bugs, report issues**: Help advance development of your project by reporting issues you encounter. Just tap „Submit“ and enter the details.
* **Work in groups**: Groups help you and your team organize your work. See your existing groups or apply for new ones.
* **Announcing: Announcements**: Announcements are messages by admins or moderators to help you quickly find out about new bugs and issues.
* **Show off your personality**: Quickly change themes or tint colors to create the experience you want. Your settings are synchronized via the Cloud.
* **Night and Day. Yours to choose**: Timed Theme allows you to automatically apply the Dark theme when it's night. And when it's day, the theme will be Light again.
* **With Updates, it keeps getting better**: Every release of Bug Reporter⁷ now comes with Software Update to keep your app always up to date, because Updates are downloaded automatically.

* NOTE: Groups and Announcements currently not available in Demo *

Software Update
===============
Introducing Software Update, a new method to update the app's offline cache.
Available in "Settings", Software Update searches for any changes to the cached files (```cache.php```) or update files (```update.json```) and applies them on demand or when the app is restarted.
Software Update automatically searches for updates on launch and informs users when an update is available.

Software Update requires working Application Cache. Current approach relies on PHP, so this requires a PHP capable server.

Changelog
=========
See [CHANGELOG.md](https://github.com/SniperGER/Bug-Reporter7/blob/master/CHANGELOG.md)
Copyright
=========
© 2014 Janik Schmidt, All rights reserved.
Permission to use this App as an example for Framework7 is hereby granted.

Bug Reporter7 uses Framework7, a full featured HTML Framework for building iOS 7 apps, created by Vladimir Kharlampidi.
Additionally, code from the example app ToDo7 has been used to enable Offline Mode and Cache Updating.

CHANGELOG.md
# Changelog
Release 1.0.2-1158 Update 1 (August 16)
---------------------------
* [IMPROVE] Changed way "Tips" section is closed  
  *NOTE: "Tips" doesn't navigate back currently as there are some technical issues when manually navigating back*


Release 1.0.2-1558 (August 16)
------------------
* [ADD] Preparation for Localization Update (1.0.3)
* [ADD] Preparation for Framework7 Update (0.9.5)
* [ADD] Added OTA Software Update
* [ADD] Added update progress screen
* [ADD] Added section to view previously installed updates
* [ADD] Added launch images for 3.5" and 4" devices
* [FIX] Fixed the background image in Bugs view (image is temporary, will be replaced soon)
* [IMPROVE] Restructured HTML File layout and Settings menu

Release 1.0.1-2341 (August 13)
------------------
* [ADD] Added option to remove all website data
* [ADD] Background Image in Bug view (Open, closed and Archived; only visible when no bugs are loaded)
* [FIX] Fixed Deadlock after completing "Tips" section
* [IMPROVE] Restructured Bug information layout (Dates FTW!)  
  *NOTE: If you have created bugs before, you need to reset them through Settings > "Reset All Website Data" or you'll have some "undefined" texts.*
* [IMPROVE] Removed unnecessary code lines (still doesn't reduce size that much)

Release 1.0-2213 (August 12)
----------------
* Initial Release


cache.php
<?php
header('Content-Type: text/cache-manifest');
$filesToCache = array(
    './index.html',
    
    './css/framework7.css', 
    './css/framework7.themes.css', 
    './css/bugreporter7.css',
    
    './frames/bugs/bugs-archived.html', 
    './frames/bugs/bugs-closed.html', 
    './frames/bugs/bugs-open.html', 
    './frames/getting-started.html',
    './frames/settings/index.html',
    './frames/settings/update/index.html',
    './frames/settings/design/theme.html',
    './frames/settings/design/tint.html',

    './img/demo_groups.png',
    './img/demo_notifications.png', 
    './img/demo_overview.png', 
    './img/demo_submit.png', 
    './img/demo_themes.png', 
    './img/demo_timed-theme.png', 
    './img/demo_universal.png',
    './img/bug_1.png',
    './img/bug_1@2x.png',

    './js/br7.common.js',
    './js/br7.handlers.js',
    './js/br7.functions.js',
    './js/cookies.js',
    './js/framework7.min.js',
    './js/jquery.js',
    './js/md5.js',
    
    './locales/de/translation.json',
    './locales/de-DE/translation.json'
);
?>
CACHE MANIFEST

CACHE:
<?php
// Print files that we need to cache and store hash data
$hashes = '';
foreach($filesToCache as $file) {
    echo $file."\n";
    $hashes.=md5_file($file);
};
?>

NETWORK:
*
update.json

EXTERNAL:
update.json

# Hash Version: <?=md5($hashes)?>
# Update Hash: <?=md5_file("./update.json")?>

js/br7.js
var br7 = new Framework7({
	animateNavBackIcon: true,
	cache: false,
	cacheDuration: 1000,
	sortable: false,
	modalTitle: unescape("Bug Reporter%u2077")
});
var openBugsCount = 0;
$(document).ready(function() {
	login_checkCookie();
	rebindUserPanel();
	if (!localStorage.doneTutorial) {
		br7.popup(".popup-getting-started");
	}
});

var timeouts = []

var statusbarTheme = br7.formGetData("form-theme") ? $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))).theme : "default";

if (window.navigator.standalone && statusbarTheme != "dark") {
	$("meta[name='apple-mobile-web-app-status-bar-style']").removeAttr("content");
}


var bugsStorage = localStorage.bugsData ? JSON.parse(localStorage.bugsData) : [];

var $$ = Framework7.$;

var mainView = br7.addView('.view-main', {
	dynamicNavbar: true
});
var loginView = br7.addView('.view-login', {
	dynamicNavbar: true
});
var gettingStartedView = br7.addView('.view-getting-started', {
	swipeBackPages: false
});

var templateHTML;


br7.onPageInit('settings', function () {
	$('form#form-theme li').on("click", function() {
		setTimeout(function() {
			changeDesign("theme");
		}, 10);
	});
	$('form#form-tint li').on("click", function() {
		setTimeout(function() {
			changeDesign("tint");
		}, 10);
	});
	
	$$('.save-button').on("click", function() {
		br7.alert("Error: The SQL database could not be contacted. Form data has been saved locally.","SQL Database Error");
	});

	$$('.reset-local-storage').on("click", function() {
		var clickedLink = $(this);
		if (br7.device.iphone) {
			var buttons1 = [
				{
					text: "This will remove data that is used to display Bugs and Settings. Installed updates are not affected. The App will be restarted once the process is finished.",
					label: true
				},
				{
					text: "Reset Data",
					red: true,
					onClick: function() {
						clearInterval(timedTheme);
						localStorage.clear();
						window.location.reload();
					}
				}
			];
			var buttons2 = [
				{
					text: "Cancel",
					bold: true
				}
			];
			var group = [buttons1, buttons2];
			br7.actions(group);
		
		} else {
			br7.modal({
				title: "Remove Website Data",
				text: "This will remove data that is used to display Bugs and Settings. Installed updates are not affected. The App will be restarted once the process is finished.",
				buttons: [
					{
						text: "Reset Data",
						onClick: function() {
							clearInterval(timedTheme);
							localStorage.clear();
							window.location.reload();
						}
					},
					{
						text: "Cancel"
					}
				]
			});

		}
	});

});

br7.onPageInit("getting-started-main", function() {
	var mySlider = br7.slider('.slider-container-h', {
			spaceBetween: 100,
			pagination:'.slider-pagination-h',
			paginationHide: false
	});
	var vSlider = br7.slider('.slider-container-v', {
			spaceBetween: 100,
			pagination:'.slider-pagination-v',
			paginationHide: false
	});
	$('[data-page="getting-started-main"] .close-popup').on("click", function() {
		setTimeout(function() {
			//gettingStartedView.goBack();
		}, 300);
	});
});

br7.onPageAfterAnimation("register", function() {
	$$('.reg-button').on("click", function() {
		login_demo();
	});
});

br7.onPageBeforeAnimation("index", function() {
	login_checkCookie();
	openBugsCount = 0;
	for (var i = 0; i < bugsStorage.length; i++) {
		var bugsItem = bugsStorage[i];
		if (bugsItem.status == "open") {
			openBugsCount++;
		}
	}
	$('#open-badge').html(openBugsCount);
});

br7.onPageBeforeAnimation("bugs-open", function() {
	buildHTML("open");
	br7 = new Framework7();
});
br7.onPageBeforeAnimation("bugs-closed", function() {
	buildHTML("closed");
	br7 = new Framework7();
});
br7.onPageBeforeAnimation("bugs-archived", function() {
	buildHTML("archived");
	br7 = new Framework7();
});

$$('.signin-button').on("click", function() {
	login_demo();
});

$$('.register-button').on("click", function() {
	br7.alert("test");
});

$$('.logout-button').on("click", function() {
	eraseCookie("username");
	$('#user-label').html("User is not signed in");
});

$$('.erase-cookies').on("click", function() {
	eraseCookie("username");
});

$$('.close-popup').on('click', function() {
	$(window).trigger("resize");
});


$('.submit-bug').on("click", function() {
	if (readCookie("username") && $('[name="newBugName"]').val() != "" && $('[name="newBugDesc"]').val() != "") {
		var date = new Date();
		var monthNames = [ "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec" ];
		var dateString = ("0" + date.getDate()).slice(-2) + "-" + monthNames[date.getMonth()] + "-" + date.getFullYear() + " " +  ("0" + date.getHours()).slice(-2) + ":" + ("0" + date.getMinutes()).slice(-2);
		bugsStorage.push({
			title: $('[name="newBugName"]').val() + " ",
			category: $('[name="newBugCat"]').val() + " ",
			description: $('[name="newBugDesc"]').val() + " ",
			author: readCookie("username"),
			status: "open",
			id: bugsStorage.length +1,
			date: dateString
		});		localStorage.bugsData = JSON.stringify(bugsStorage);
		br7.showPreloader("Submitting...");
		setTimeout(function() {
			br7.hidePreloader();
			setTimeout(function() {
				br7.closeModal();
				buildHTML("open");
				setTimeout(function() {
					$('[name="newBugName"]').val("");
					$('[name="newBugDesc"]').val("");
				}, 300);
			}, 500);
		}, 2000);
	} else if (!readCookie("username")) {
		br7.alert("This feature requires a valid login.","User not logged in");
	} else if ($('[name="newBugName"]').val() == "") {
		br7.alert("Please specify a title before continuing.", "Missing fields");
	} else if ($('[name="newBugDesc"]').val() == "") {
		br7.alert("Please enter some details before continuing.", "Missing fields");
	}
});


function login() {
	loginView.loadPage("frames/login.php?u="+$('input#username').val()+"&p="+MD5($('input#password').val()));
}
function login_demo() {
	br7.showPreloader("Signing in");
	setTimeout(function() {
		br7.hidePreloader();
		createCookie("username","Demo User",24);
	}, 2500);
	setTimeout(function() {
		br7.closeModal(".popup-login");
		if (readCookie("username") == undefined) {
			$('#user-label').html("Login failed. Please allow Cookies on this website.");
		} else {
			$('#user-label').html(readCookie("username"));
		}
		loginView.goBack(undefined, false); // URL: undefined, animatePages: false
	}, 3000);
}
function login_checkCookie() {
	if (!readCookie("username")) {
		//br7.popup(".popup-login");
	} else {
		$('#user-label').html(readCookie("username"));
	}
}
function rebindUserPanel() {
	$$('#user-label').on("click", function() {
		var clickedLink = this;
		var buttons = [
		{
				text: 'Sign off',
				bold: true,
				onClick: function () {
					eraseCookie("username");
			$('#user-label').html("User is not signed in");
				}
		}];
		var buttons2 = [
		{
				text: 'Cancel',
				red: true,
				onClick: function () {
				}
		}];
	
		if (readCookie("username")) {
			if (br7.device.iphone) {
				br7.actions([buttons, buttons2]);
			} else {
				br7.popover(".popover-usercp", clickedLink);
			}
		} else {
			br7.popup(".popup-login");
		}
	});
}

function changeDesign(key) {
	var settingsTheme = $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))) != null ? $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))) : "default";
	var settingsTint = $.parseJSON(JSON.stringify(br7.formGetData("form-tint"))) != null ? $.parseJSON(JSON.stringify(br7.formGetData("form-tint"))) : "blue";
	switch (key) {
		case "theme":
			$('body').attr("data-theme", settingsTheme.theme);
			if (settingsTheme.theme == "dark") {
				$("meta[name='apple-mobile-web-app-status-bar-style']").removeAttr("content");
			} else {
				$("meta[name='apple-mobile-web-app-status-bar-style']").attr("content","black-translucent");
			}
			break;
		case "tint":
			$('body').attr("data-color", settingsTint.tint);
			break;
		case "init":
			$('body').attr("data-theme", settingsTheme.theme);
			$('body').attr("data-color", settingsTint.tint);
			break;
		default: break;
	}
}
changeDesign("init");

function buildHTML(page) {
	var bugTemplate;
	switch (page) {
		case "open": bugTemplate = $$('#bug-open-template').html(); break;
		case "closed": bugTemplate = $$('#bug-closed-template').html(); break;
		case "archived": bugTemplate = $$('#bug-archived-template').html(); break;
		default: break;
	}
	var html = '';
	for (var i = 0; i < bugsStorage.length; i++) {
		var bugsItem = bugsStorage[i];
		if (bugsItem.status == page) {
			html += bugTemplate.replace(/{{title}}/g, bugsItem.title).replace(/{{category}}/g, bugsItem.category).replace(/{{description}}/g, bugsItem.description).replace(/{{author}}/g, bugsItem.author).replace(/{{status}}/g, bugsItem.status).replace(/{{id}}/g, bugsItem.id).replace(/{{date}}/g, bugsItem.date);
		}
		switch (page) {
			case "open":
				$('#bugs-open-content').html(html);
				if ($('#bugs-open-content').html() != "") {
					$('.page-content.bug-bg').removeClass("bug-bg");
				}
				$$('.action1').on("click", function() {
					var issueID = parseInt($(this).closest(".swipeout").children(".swipeout-content").children(".issueID").html());
					var bugObject = bugsStorage[issueID-1];
					bugObject.status = "archived";
					bugsStorage[issueID-1] = bugObject;
					localStorage.bugsData = JSON.stringify(bugsStorage);
					var obj = $(this);
					timeouts.push(setTimeout(function() {
						$(obj).closest("div.list-block").remove();
						if ($('#bugs-open-content').children("div.list-block").length < 1) {
							$('.page-content.bug-no-bg').addClass("bug-bg");
						}
					}, 300));
				});
				$$('.action2').on("click", function() {
					var issueID = parseInt($(this).closest(".swipeout").children(".swipeout-content").children(".issueID").html());
					var bugObject = bugsStorage[issueID-1];
					bugObject.status = "closed";
					bugsStorage[issueID-1] = bugObject;
					localStorage.bugsData = JSON.stringify(bugsStorage);
					var obj = $(this);
					timeouts.push(setTimeout(function() {
						$(obj).closest("div.list-block").remove();
						if ($('#bugs-open-content').children("div.list-block").length < 1) {
							$('.page-content.bug-no-bg').addClass("bug-bg");
						}
					}, 300));
				});
				break;
			case "closed":
				if ($('#bugs-closed-content').html() != "") {
					$('.page-content.bug-bg').removeClass("bug-bg");
				}
				$('#bugs-closed-content').html(html);
				$$('.action1').on("click", function() {
					var issueID = parseInt($(this).closest(".swipeout").children(".swipeout-content").children(".issueID").html());
					var bugObject = bugsStorage[issueID-1];
					bugObject.status = "archived";
					bugsStorage[issueID-1] = bugObject;
					localStorage.bugsData = JSON.stringify(bugsStorage);
					var obj = $(this);
					timeouts.push(setTimeout(function() {
						$(obj).closest("div.list-block").remove();
						if ($('#bugs-closed-content').children("div.list-block").length < 1) {
							$('.page-content.bug-no-bg').addClass("bug-bg");
						}
					}, 300));
				});
				$$('.action2').on("click", function() {
					var issueID = parseInt($(this).closest(".swipeout").children(".swipeout-content").children(".issueID").html());
					var bugObject = bugsStorage[issueID-1];
					bugObject.status = "open";
					bugsStorage[issueID-1] = bugObject;
					localStorage.bugsData = JSON.stringify(bugsStorage);
					var obj = $(this);
					timeouts.push(setTimeout(function() {
						$(obj).closest("div.list-block").remove();
						if ($('#bugs-closed-content').children("div.list-block").length < 1) {
							$('.page-content.bug-no-bg').addClass("bug-bg");
						}
					}, 300));
				});
				break;
			case "archived":
				if ($('#bugs-closed-content').html() != "") {
					$('.page-content.bug-bg').removeClass("bug-bg");
				}
				$('#bugs-archived-content').html(html);
				$$('.action1').on("click", function() {
					var issueID = parseInt($(this).closest(".swipeout").children(".swipeout-content").children(".issueID").html());
					var bugObject = bugsStorage[issueID-1];
					var index = bugsStorage.indexOf(bugObject);
					var obj = $(this);
					timeouts.push(setTimeout(function() {
						$(obj).closest("div.list-block").remove();
						if ($('#bugs-archived-content').children("div.list-block").length < 1) {
							$('.page-content.bug-no-bg').addClass("bug-bg");
						}
					}, 300));
					if (index > -1) {
						bugsStorage.splice(index, 1);
						localStorage.bugsData = JSON.stringify(bugsStorage);
					}
				});
				break;
			default: break;
		}
	}
}

var timedTheme = setInterval(function() {
	var timedThemeSwitch = br7.formGetData("form-theme") != undefined ? $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))).switchTimedTheme : []
	if (timedThemeSwitch.length != 0) {
		var time = new Date();
		var hours = time.getHours();
		
		$('ul#themes').parent().addClass("disabled");
		
		if (hours >= 22 || hours <= 7) {
			$('body').attr("data-theme", "dark");
			br7.formStoreData("form-theme", {
				"theme":"dark",
				"switchTimedTheme": $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))).switchTimedTheme
			});
			br7.formFromJSON('#form-theme', {
				"theme":"dark",
				"switchTimedTheme": $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))).switchTimedTheme
			});
		} else {
			$('body').attr("data-theme", "default");
			br7.formStoreData("form-theme", {
				"theme":"default",
				"switchTimedTheme": $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))).switchTimedTheme
			});
			br7.formFromJSON('#form-theme', {
				"theme":"default",
				"switchTimedTheme": $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))).switchTimedTheme
			});
		}
	} else {
		$('ul#themes').parent().removeClass("disabled");
	}
}, 100);

window.addEventListener('load', function (e) {
    window.applicationCache.addEventListener('updateready', function (e) {
        if (window.applicationCache.status === window.applicationCache.UPDATEREADY) {
            // Browser downloaded a new app cache.
            br7.confirm(unescape('Bug Reporter%u2077 has been updated. Would you like to install this update?'), function () {
                window.location.reload();
            });
        }
    }, false);
}, false);

for (var i = 0; i < bugsStorage.length; i++) {
	var bugsItem = bugsStorage[i];
	if (bugsItem.status == "open") {
		openBugsCount++;
	}
}
$('#open-badge').html(openBugsCount);

js/br7.functions.js
function login() {
	loginView.loadPage("frames/login/login.php?u="+$('input#username').val()+"&p="+MD5($('input#password').val()));
}
function login_demo() {
	br7.showPreloader("Signing in");
	setTimeout(function() {
		br7.hidePreloader();
		createCookie("username","Demo User",24);
	}, 2500);
	setTimeout(function() {
		br7.closeModal(".popup-login");
		if (readCookie("username") == undefined) {
			$('#user-label').html("Login failed. Please allow Cookies on this website.");
		} else {
			$('#user-label').html(readCookie("username")).removeAttr("data-i18n");
		}
		loginView.goBack(undefined, false); // URL: undefined, animatePages: false
	}, 3000);
}
function login_checkCookie() {
	if (!readCookie("username")) {
		//br7.popup(".popup-login");
	} else {
		$('#user-label').html(readCookie("username")).removeAttr("data-i18n");
	}
}
function rebindUserPanel() {
	$('#user-label').off("click");
	$('#user-label').on("click", function() {
		var clickedLink = this;
		var buttons = [
		{
				text: 'Sign off',
				bold: true,
				onClick: function () {
					eraseCookie("username");
					$('#user-label').attr("data-i18n","NOT_SIGNED_IN").html("User is not signed in");
/* 					$('body').i18n(); */
				}
		}];
		var buttons2 = [
		{
				text: 'Cancel',
				red: true,
				onClick: function () {
				}
		}];
	
		if (readCookie("username")) {
			if (br7.device.iphone) {
				br7.actions([buttons, buttons2]);
			} else {
				br7.popover(".popover-usercp", clickedLink);
			}
		} else {
			br7.popup(".popup-login");
		}
	});
}

function br7update() {
	var html= "<div class='animate-height'> \
	<div class='content-block tablet-inset'> \
	    <div class='content-block-inner'> \
	    	<img src='iTunesArtwork.png' style='width: 60px; height: 60px; vertical-align: top; float: left;'> \
	    	<p style='margin: 0; margin-left: 8px;'> \
	    		<span style='font-weight: 500; font-size: 16px; margin-left: 8px'>Bug Reporter&#8311; {{version}}<br></span> \
	    		<span style='font-size: 14px; margin-left: 8px'>Sniper_GER<br></span> \
	    		<span style='font-size: 14px; margin-left: 8px'>Downloaded</span> \
	    	</p> \
	    	<p>{{description}}<br><br>For more information, visit:<br><a href='https://github.com/SniperGER/Bug-Reporter7' class='external' target='_blank' style='text-decoration: underline;'>https://github.com/SniperGER/Bug-Reporter7</a>{{instructions}}</p> \
	    </div> \
	</div> \
	</div> \
	<div class='list-block tablet-inset'> \
	    <ul> \
	    	<li> \
	    		<a href='#' class='item-link open-popup' data-popup='.popup-update-detail'> \
	    			<div class='item-content'> \
	    				<div class='item-inner'> \
	    					<div class='item-title'>Details</div> \
	    				</div> \
	    			</div> \
	    		</a> \
	    	</li> \
	    </ul> \
	</div> \
	<div class='list-block tablet-inset'> \
	    <ul> \
	    	<li class='center item-button'> \
	    		<a href='#' class='update-button'> \
	    			<div class='item-content'> \
	    				<div class='item-inner'> \
	    					<div class='item-title' style='width: 100%;'>Install</div> \
	    				</div> \
	    			</div> \
	    		</a> \
	    	</li> \
	    </ul> \
	</div>";
	$.getJSON("update.json", function(data) {
	    $('p#update-status').closest(".page-content").html(html.replace(/{{version}}/g, data.en.updates[0].version).replace(/{{description}}/g, data.en.updates[0].description).replace(/{{instructions}}/g, ""));
		$('span#content').html(data.en.updates[0].content);
	    $('.update-button').on("click", function() {
	    	br7.modal({
	    		title: "Software Update",
	    		text: unescape("Bug Reporter%u2077 "+data.en.updates[0].version+" will begin installing. The app will restart when installation is finished."),
	    		buttons: [
	    			{
	    				text: "Later",
	    				onClick: function() {
		    				br7.closeModal(); 
	    				}
	    			},
	    			{
	    				text: "Install",
	    				bold: true,
	    				onClick: function() {
	    					br7.hidePreloader();
	    					$('body').append("<div class=\"update-view\"><img src=\"iTunesArtwork.png\" /></div>");
	    					setTimeout(function() {
	    						$('.update-view').append("<div class=\"progress-bar\"><div class=\"inner-progress\" id=\"update-progress\"></div></div>");
	    						interval = setInterval(function() {
	    							addProgress("update-progress");
	    						}, Math.random() * 750);
	    					}, 1000);
	    				}
	    			}
	    		]
	    	});
	    });
	});
}

function changeDesign(key) {
	var settingsTheme = $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))) != null ? $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))) : "default";
	var settingsTint = $.parseJSON(JSON.stringify(br7.formGetData("form-tint"))) != null ? $.parseJSON(JSON.stringify(br7.formGetData("form-tint"))) : "blue";
	switch (key) {
		case "theme":
			$('body').removeClass("layout-dark layout-white").addClass("layout-"+settingsTheme.theme).attr("data-theme", settingsTheme.theme);
			if (settingsTheme.theme == "dark") {
				$("meta[name='apple-mobile-web-app-status-bar-style']").removeAttr("content");
			} else {
				$("meta[name='apple-mobile-web-app-status-bar-style']").attr("content","black-translucent");
			}
			break;
		case "tint":
			$('body').removeClass("theme-white theme-black theme-yellow theme-red theme-blue theme-green theme-pink theme-lightblue theme-orange theme-gray").addClass("theme-"+settingsTint.tint).attr("data-color", settingsTint.tint);
			break;
		case "init":
			$('body').removeClass("layout-dark layout-white").addClass("layout-"+settingsTheme.theme).attr("data-theme", settingsTheme.theme);
			$('body').removeClass("theme-white theme-black theme-yellow theme-red theme-blue theme-green theme-pink theme-lightblue theme-orange theme-gray").addClass("theme-"+settingsTint.tint).attr("data-color", settingsTint.tint);
			break;
		default: break;
	}
}
changeDesign("init");

var progress = 0;
function giveRandom(limit) {
	return Math.floor(Math.random() * limit) + 1;
}
function addProgress(selector, callback) {
	if (progress < 100) {
		var randomValue = giveRandom(30);
		if ((progress + randomValue) <= 100) {
			progress = progress + randomValue;
		} else {
			progress = 100;
		}
	} else if (progress >=100) {
		progress = 100;
		clearInterval(interval);
		interval = undefined;
		setTimeout(function() { $('.progress-bar').remove() }, 1000);
		setTimeout(function() { window.location.reload() }, 2000);

	}
	$('#'+selector).css("width",progress+"%");
}

function buildHTML(page) {
	var bugTemplate;
	switch (page) {
		case "open": bugTemplate = $$('#bug-open-template').html(); break;
		case "closed": bugTemplate = $$('#bug-closed-template').html(); break;
		case "archived": bugTemplate = $$('#bug-archived-template').html(); break;
		default: break;
	}
	var html = '';
	for (var i = 0; i < bugsStorage.length; i++) {
		var bugsItem = bugsStorage[i];
		if (bugsItem.status == page) {
			html += bugTemplate.replace(/{{title}}/g, bugsItem.title).replace(/{{category}}/g, bugsItem.category).replace(/{{description}}/g, bugsItem.description).replace(/{{author}}/g, bugsItem.author).replace(/{{status}}/g, bugsItem.status).replace(/{{id}}/g, bugsItem.id).replace(/{{date}}/g, bugsItem.date);
		}
		switch (page) {
			case "open":
				$('#bugs-open-content').html(html);
				if ($('#bugs-open-content').children("div.list-block").length > 0) {
					$('.page-content.bug-bg#bugs-open-content').removeClass("bug-bg");
				}
				$$('.action1').on("click", function() {
					var issueID = parseInt($(this).closest(".swipeout").children(".swipeout-content").children(".issueID").html());
					var bugObject = bugsStorage[issueID-1];
					bugObject.status = "archived";
					bugsStorage[issueID-1] = bugObject;
					localStorage.bugsData = JSON.stringify(bugsStorage);
					var obj = $(this);
					timeouts.push(setTimeout(function() {
						$(obj).closest("div.list-block").remove();
						if ($('#bugs-open-content').children("div.list-block").length < 1) {
							$('.page-content.bug-no-bg').addClass("bug-bg").html("");
						}
					}, 300));
				});
				$$('.action2').on("click", function() {
					var issueID = parseInt($(this).closest(".swipeout").children(".swipeout-content").children(".issueID").html());
					var bugObject = bugsStorage[issueID-1];
					bugObject.status = "closed";
					bugsStorage[issueID-1] = bugObject;
					localStorage.bugsData = JSON.stringify(bugsStorage);
					var obj = $(this);
					timeouts.push(setTimeout(function() {
						$(obj).closest("div.list-block").remove();
						if ($('#bugs-open-content').children("div.list-block").length < 1) {
							$('.page-content.bug-no-bg').addClass("bug-bg").html("");
						}
					}, 300));
				});
				break;
			case "closed":
				$('#bugs-closed-content').html(html);
				if ($('#bugs-closed-content').children("div.list-block").length > 0) {
					$('.page-content.bug-bg#bugs-closed-content').removeClass("bug-bg");
				}
				$$('.action1').on("click", function() {
					var issueID = parseInt($(this).closest(".swipeout").children(".swipeout-content").children(".issueID").html());
					var bugObject = bugsStorage[issueID-1];
					bugObject.status = "archived";
					bugsStorage[issueID-1] = bugObject;
					localStorage.bugsData = JSON.stringify(bugsStorage);
					var obj = $(this);
					timeouts.push(setTimeout(function() {
						$(obj).closest("div.list-block").remove();
						if ($('#bugs-closed-content').children("div.list-block").length < 1) {
							$('.page-content.bug-no-bg').addClass("bug-bg").html("");
						}
					}, 300));
				});
				$$('.action2').on("click", function() {
					var issueID = parseInt($(this).closest(".swipeout").children(".swipeout-content").children(".issueID").html());
					var bugObject = bugsStorage[issueID-1];
					bugObject.status = "open";
					bugsStorage[issueID-1] = bugObject;
					localStorage.bugsData = JSON.stringify(bugsStorage);
					var obj = $(this);
					timeouts.push(setTimeout(function() {
						$(obj).closest("div.list-block").remove();
						if ($('#bugs-closed-content').children("div.list-block").length < 1) {
							$('.page-content.bug-no-bg').addClass("bug-bg").html("");
						}
					}, 300));
				});
				break;
			case "archived":
				$('#bugs-archived-content').html(html);
				if ($('#bugs-archived-content').children("div.list-block").length > 0) {
					$('.page-content.bug-bg#bugs-archived-content').removeClass("bug-bg");
				}
				$$('.action1').on("click", function() {
					var issueID = parseInt($(this).closest(".swipeout").children(".swipeout-content").children(".issueID").html());
					var bugObject = bugsStorage[issueID-1];
					var index = bugsStorage.indexOf(bugObject);
					var obj = $(this);
					timeouts.push(setTimeout(function() {
						$(obj).closest("div.list-block").remove();
						if ($('#bugs-archived-content').children("div.list-block").length < 1) {
							$('.page-content.bug-no-bg').addClass("bug-bg").html("");
						}
					}, 300));
					if (index > -1) {
						bugsStorage.splice(index, 1);
						localStorage.bugsData = JSON.stringify(bugsStorage);
					}
				});
				break;
			default: break;
		}
	}
}

var timedTheme = setInterval(function() {
	var timedThemeSwitch = br7.formGetData("form-theme") != undefined ? $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))).switchTimedTheme : []
	if (timedThemeSwitch.length != 0) {
		var time = new Date();
		var hours = time.getHours();
		
		$('ul#themes').parent().addClass("disabled");
		
		if (hours >= 22 || hours <= 7) {
			$('body').removeClass("layout-dark layout-white").addClass("layout-dark").attr("data-theme", "dark");
			br7.formStoreData("form-theme", {
				"theme":"dark",
				"switchTimedTheme": $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))).switchTimedTheme
			});
			br7.formFromJSON('#form-theme', {
				"theme":"dark",
				"switchTimedTheme": $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))).switchTimedTheme
			});
		} else {
			$('body').removeClass("layout-dark").attr("data-theme", "default");
			br7.formStoreData("form-theme", {
				"theme":"default",
				"switchTimedTheme": $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))).switchTimedTheme
			});
			br7.formFromJSON('#form-theme', {
				"theme":"default",
				"switchTimedTheme": $.parseJSON(JSON.stringify(br7.formGetData("form-theme"))).switchTimedTheme
			});
		}
	} else {
		$('ul#themes').parent().removeClass("disabled");
	}
}, 100);

function checkForUpdatesOnLoad() {
	if (window.applicationCache.status === window.applicationCache.UPDATEREADY) {
		$.getJSON("update.json", function(data) {
			br7.modal({
				title: "Software Update",
				text: "Version "+data.en.updates[0].version+" is now available.",
				buttons: [
					{
						text: "Close"
					},
					{
						text: "Details",
						bold: true,
						onClick: function() {
							br7.closeModal();
							mainView.loadPage("frames/settings/update");
						}
					}
				]
			});
		});
	}
}
window.addEventListener('load', function (e) {
    window.applicationCache.addEventListener('updateready', checkForUpdatesOnLoad, false);
}, false);